accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    by a list, set or map literal. It is a potentially constant element if the
+    expression is potentially constant expression.
+
+*   An `ifElement` is constant if its condition is a constant expression
+    evaluating to a Boolean value and either:
+
+    *   If the condition evaluates to `true`, the then element is a constant
+        expression, and any else element is a potentially constant element.
+
+    *   If the condition evaluates to `false`, then the then element is a
+        potentially constant element and any else element is a constant element.
+
+*   A `forElement` is never a constant element. `for` cannot be used in constant
+    set or map literals.
+
+*   It is a compile-time error to have duplicate values in a set literal or
+    duplicate keys in a map literal, according to behavior specified in the
+    runtime semantics.
+
+*   When evaluating a const collection literal, if there is an existing const
+    list, map, or set literal that contains the same series of values or entries
+    in the same order, then that object is used instead of producing a new one.
+    *In other words, constants are canonicalized.* Note that maps are only
+    canonicalized if they contain the same keys *in the same order*.
+
+### Constant equality safe
+
+A value is constant equality safe iff:
+
+*   It is an instance of a class that does not override `Object.operator==`, or
+
+*   it is an instance of `int` or `String`, or
+
+*   it is an instance of `Symbol` originally created by a symbol literal or a
+    constant invocation of the `Symbol` constructor, or
+
+*   it is an instance of `Type` originally created by evaluating a type literal
+    expression (an identifier or qualified identifier denoting a class, mixin or
+    type alias declaration, that is evaluated as an expression).
+
+All constant equality safe values are also valid constant values.
+
+## Dynamic Semantics
+
+Spread, `if`, and `for` behave similarly across all collection types. To
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a `result` sequence of
+either values or key-value pair map entries. Then a final step handles those
+appropropriately for the given collection type.
+
+### To evaluate a collection `element`:
+
+1.  If `element` is an expression element:
+
+    1.  Evaluate the element's expression and append it to `result`.
+
+1.  Else, `element` is a `mapEntry` `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value `key`.
+
+    1.  Evaluate `valueExpression` to a value `value`.
+
+    1.  Append `key: value` to `result`.
+
+1.  Else, if `element` is a spread element:
+
+    1.  Evaluate the spread expression to a value `spread`.
+
+    1.  If `entry` is null-aware and `spread` is null, continue to the next
+        element in the literal.
+
+    1.  If the collection is a map, evaluate `spread.entries.iterator` to a
```

Comment:
Do we really want to specify this for maps in terms of the entries iterator?  That's likely to be fairly expensive, since you have to allocate all of the key/value pairs.  Perhaps better to specify this for maps in terms of iterating over the keys directly?

Reply:
This will be a bigger change. I'm going to try to finish replying to all of the other smaller feedback and upload a new commit. Then I'll circle back to this.

Reply:
I did some benchmarking and wrote up a doc. After discussing this at the language meeting, we've reached some decisions. I've updated the proposal to match that.